### DISCUSSION TITLE
Python: Manipulate FunctionCallingStepwisePlanner in the interim

### DISCUSSION THREAD

**Discussion by jungwookim** [microsoft/semantic-kernel#6902] [Q&A] [answered]
Hi, Community!

I used to use Semantic Kernel 0.5.x and recently upgraded to version 1.1.1. I'm now getting accustomed to the FunctionCallingStepwisePlanner.

My use case involves a scholarship application process. I've added an AssistantPlugin with two functions: (1) Retrieve available scholarships based on provided information, and (2) Apply for a scholarship.

The issue arises when retrieving a scholarship. The response includes a content format like `Information you need to fill out: <your_name>/<your_student_number>. For example, get me "John / 123456".`. Consequently, the planner interprets `John/123456` as an argument for the ApplyScholarship function.

In other words, when I input `I'd like to apply for a scholarship.`, I expect the assistant to retrieve relevant information and ask for my personal details. However, the actual response is `Alright. You've successfully applied.`

I'm unsure how to handle this case. Could filters be helpful here, or does anyone have similar experiences or suggestions for resolving this issue?

**Comment by moonbox3** [comment, marked as answer]
Hi @jungwookim, thank you for your question. To better assist you, I have a question for you to help make sure I understand the full picture:

In your `Apply Scholarship` function, are you making the function params (ideally with Annotations) required? By making these parameters required, we're able to send that information to the model so it knows before it can invoke a function that it needs certain information. 

For example, the following kernel function:

```python
class ScholarshipPlugin:
    @kernel_function(name="apply_for_scholarship", description="Apply for a scholarship.")
    def apply_for_scholarship(
        self,
        name: Annotated[str, "The name of the applicant."],
        age: Annotated[int, "The age of the applicant."],
        scholarship: Annotated[str, "The scholarship to apply for."],
    ):
        """Apply for a scholarship."""
        return f"{name} has applied for the {scholarship} scholarship."
```

produces the following JSON schema that is sent to the model:

```json
{
    "type": "function",
    "function":
    {
        "name": "scholarship-apply_for_scholarship",
        "description": "Apply for a scholarship.",
        "parameters":
        {
            "type": "object",
            "properties":
            {
                "name":
                {
                    "type": "string",
                    "description": "The name of the applicant."
                },
                "age":
                {
                    "type": "integer",
                    "description": "The age of the applicant."
                },
                "scholarship":
                {
                    "type": "string",
                    "description": "The scholarship to apply for."
                }
            },
            "required":
            [
                "name",
                "age",
                "scholarship"
            ]
        }
    }
}
```

And using vanilla function calling, I have this interaction with the model:

```text
User:> I'd like to apply for a scholarship
Mosscap:> Ah, the pursuit of knowledge and opportunity! I would be delighted to assist you with that. Could you kindly provide me with the following details?

1. Your full name.
2. Your age.
3. The specific scholarship you are interested in applying for.

Once I have this information, I can proceed with the application for you.
```

This works well when it's just function calling; however, while using the FunctionCallingStepwisePlanner, as you know, it creates a plan and then executes it with no further input. Thus, you could use our concept of `filters`, yes, to help with this scenario. You'd want to have a look at the following [concept sample](https://github.com/microsoft/semantic-kernel/blob/main/python/samples/concepts/filtering/function_invocation_filters.py).

**Reply by jungwookim** [reply]
@moonbox3 Thank you for your answer. Yes, you get my point and I wrote a plugin with Annotations.

I have been experimenting two approaches. One is function calling and the other one is FunctionCallingStepwisePlanner. I find out the function calling works but FunctionCallingStepwisePlanner. That's the reason why I was asking a question.

Alright then, I would obviously use `filters` if needed.

By the way, any debug tips how plugin-function would be generated as yaml like your example? Because I might need to handle more variations such as `array object` and `optional fields` in plugin-function.
